File: tests/cpp/test_vtp.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// System includes
#include <filesystem>
#include <string>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vtp.hpp"

namespace {

void rt(const mt::Mesh& mesh, bool binary, bool zlib) {
    mt::roundtrip([=](const std::string& p,
                      const mt::Mesh& m) { meshioplusplus::write_vtp(p, m, binary, zlib); },
                  [](const std::string& p) { return meshioplusplus::read_vtp(p); }, mesh, ".vtp");
}

mt::Mesh vertex_line_tri_mesh() {
    mt::Mesh mesh = mt::tri_mesh();
    mesh.AddCellBlock("vertex", mt::conn_from({{0}, {2}}));
    mesh.AddCellBlock("line", mt::conn_from({{0, 1}, {1, 2}}));
    return mesh;
}

// Rectangular (uniform row size) polygon block — mt::roundtrip's comparison
// requires Conn(), so the jagged case gets its own manual test below.
mt::Mesh polygon_mesh() {
    mt::Mesh mesh = mt::make_mesh({{0, 0, 0}, {1, 0, 0}, {1.5, 1, 0}, {0.5, 1.7, 0}, {-0.5, 1, 0}},
                                  "triangle", {{0, 1, 2}});
    mesh.AddCellBlock("polygon", mt::conn_from({{0, 1, 2, 3, 4}}));
    return mesh;
}

}  // namespace

TEST(Vtp, AsciiSurface) {
    rt(mt::tri_mesh(), false, false);
    rt(mt::quad_mesh(), false, false);
    rt(mt::tri_quad_mesh(), false, false);
}
TEST(Vtp, Ascii2D) {
    rt(mt::tri_mesh_2d(), false, false);
}
TEST(Vtp, AsciiVertsLines) {
    rt(vertex_line_tri_mesh(), false, false);
}
TEST(Vtp, AsciiPolygon) {
    rt(polygon_mesh(), false, false);
}
TEST(Vtp, BinaryUncompressed) {
    rt(mt::tri_mesh(), true, false);
    rt(polygon_mesh(), true, false);
}
#ifdef MESHIOPLUSPLUS_HAS_ZLIB
TEST(Vtp, BinaryZlib) {
    rt(mt::tri_mesh(), true, true);
    rt(mt::tri_quad_mesh(), true, true);
}
#endif

TEST(Vtp, RaggedPolygonWrite) {
    // A jagged polygon block writes as Polys rows; on read the 4-noded row
    // legitimately comes back as a quad (PolyData has no cell-type array).
    mt::Mesh mesh = mt::make_mesh({{0, 0, 0}, {1, 0, 0}, {1.5, 1, 0}, {0.5, 1.7, 0}, {-0.5, 1, 0}},
                                  "triangle", {{0, 1, 2}});
    mesh.AddPolygonBlock("polygon", {{0, 1, 2, 3, 4}, {0, 1, 2, 3}});
    const std::string path = mt::temp_path(".vtp");
    meshioplusplus::write_vtp(path, mesh, /*binary=*/false, /*zlib=*/false);
    mt::Mesh out = meshioplusplus::read_vtp(path);
    ASSERT_EQ(out.NumCellBlocks(), 3u);
    EXPECT_EQ(out.Cells(0).Type(), "triangle");
    EXPECT_EQ(out.Cells(1).Type(), "polygon");
    EXPECT_EQ(out.Cells(1).NumCells(), 1u);
    EXPECT_EQ(out.Cells(1).NodesPerCell(), 5u);
    EXPECT_EQ(out.Cells(2).Type(), "quad");
    EXPECT_EQ(out.Cells(2).NumCells(), 1u);
    std::error_code ec;
    std::filesystem::remove(path, ec);
}

TEST(Vtp, VolumeCellsRejected) {
    EXPECT_THROW(meshioplusplus::write_vtp(mt::temp_path(".vtp"), mt::tet_mesh(), false, false),
                 meshioplusplus::WriteError);
    EXPECT_THROW(
        meshioplusplus::write_vtp(mt::temp_path(".vtp"), mt::triangle6_mesh(), false, false),
        meshioplusplus::WriteError);
}

File: tests/python/test_vtp.py
import copy

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus.vtp import _vtp

from . import helpers

test_set = [
    helpers.line_mesh,
    helpers.tri_mesh,
    helpers.tri_mesh_2d,
    helpers.tri_mesh_one_cell,
    helpers.quad_mesh,
    helpers.tri_quad_mesh,
    helpers.polygon_mesh,
    helpers.polygon_mesh_one_cell,
    # NOT polygon2_mesh: PolyData has no cell-type array, so its 3-/4-noded
    # polygon cells legitimately read back as triangle/quad.
    helpers.add_point_data(helpers.tri_mesh, 1),
    helpers.add_point_data(helpers.tri_mesh, 3),
    helpers.add_cell_data(helpers.tri_mesh, [("a", (), np.float64)]),
    helpers.add_cell_data(helpers.tri_quad_mesh, [("a", (), np.float64)]),
]


@pytest.mark.parametrize("mesh", test_set)
@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test(mesh, binary, compression, tmp_path):
    def writer(filename, mesh):
        meshioplusplus.vtp.write(filename, mesh, binary=binary, compression=compression)

    tol = 1.0e-15 if binary else 1.0e-10
    helpers.write_read(
        tmp_path, writer, meshioplusplus.vtp.read, mesh, tol, extension=".vtp"
    )


@pytest.mark.parametrize("mesh", [helpers.tri_mesh, helpers.polygon_mesh])
@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_cross_compat(mesh, binary, compression, tmp_path):
    # shim (C++ where available) write -> pure-Python read
    p = tmp_path / "cross.vtp"
    meshioplusplus.vtp.write(
        p, copy.deepcopy(mesh), binary=binary, compression=compression
    )
    out = _vtp.read(p)
    assert np.allclose(mesh.points, out.points, atol=1.0e-13, rtol=0.0)

    # pure-Python write -> shim (C++ where available) read
    _vtp.write(p, copy.deepcopy(mesh), binary=binary, compression=compression)
    out = meshioplusplus.vtp.read(p)
    assert np.allclose(mesh.points, out.points, atol=1.0e-13, rtol=0.0)


def test_lzma_roundtrip(tmp_path):
    # lzma is Python-only (the C++ path raises and the shim falls back).
    def writer(filename, mesh):
        meshioplusplus.vtp.write(filename, mesh, binary=True, compression="lzma")

    helpers.write_read(
        tmp_path,
        writer,
        meshioplusplus.vtp.read,
        helpers.tri_mesh,
        1.0e-15,
        extension=".vtp",
    )


@pytest.mark.parametrize("writer", [meshioplusplus.vtp.write, _vtp.write])
def test_volume_cells_raise(writer, tmp_path):
    with pytest.raises(meshioplusplus.WriteError):
        writer(tmp_path / "vol.vtp", copy.deepcopy(helpers.tet_mesh))


def test_generic_io(tmp_path):
    helpers.generic_io(tmp_path / "test.vtp")
    # With additional, insignificant suffix:
    helpers.generic_io(tmp_path / "test.0.vtp")
